### PR TITLE
Bug 1208416 - Report on Hello MAUs per new MAU definition. r=mikedeboer

### PR DESCRIPTION
**Pull request by mancas**

### CODE REVIEW

File: add-on/chrome/bootstrap.js [hunk 1]
```diff
@@ -172,6 +174,46 @@ var WindowListener = {
...
       /**
+       * Method to send telemetry event when the user opens panel
+       * and hasn't been send yet.
+       */
+      _shouldSendMauTelemetry: function() {
+        this.LoopAPI.sendMessageToHandler({
+          name: "GetLoopPref",
+          data: ["mau.openPanel"]
+        }, pref => {
+          if (!pref) {
```
Comment (mikedeboer) [inline add-on/chrome/bootstrap.js]: This doesn't really address the 'one month ago' issue... What I'd recommend is to add logic to the 'TelemetryAddValue' message handler to deal with MAU events:
1. Each area of the UI that needs measuring will be dumb and call `TelemetryAddValue`.
2. `TelemetryAddValue` sees an incoming request for 'LOOP_MAU'...
3. ...so it reads the current integer value of the corresponding pref
4. Compares nowDate - prefDate >= 30days
5. If so, we're all good and telemetry value is added. The pref is set to `nowDate` after that, so we won't record that event for another 30 days at least.

What do you think?

File: add-on/preferences/prefs.js [hunk 2]
```diff
@@ -21,6 +21,11 @@ pref("loop.feedback.dateLastSeenSec", 0);
 pref("loop.feedback.periodSec", 15770000); // 6 months.
 pref("loop.feedback.formURL", "https://www.mozilla.org/firefox/hello/npssurvey/");
 pref("loop.feedback.manualFormURL", "https://www.mozilla.org/firefox/hello/feedbacksurvey/");
+pref("loop.mau.openPanel", false);
```
Comment (mikedeboer) [inline add-on/preferences/prefs.js]: So with the proposal above, these prefs would be initialized with `0`.

File: add-on/chrome/modules/MozLoopAPI.jsm [hunk 3]
```diff
@@ -1001,7 +1002,48 @@ const kMessageHandlers = {
...
+        // 2592000 seconds in 30 days
+        if (pref === 0 || delta >= 2592000 * 1000) {
+          Services.telemetry.getHistogramById(histogramId).add(value);
+          let timestamp = Math.floor(new Date().getTime() / 1000);
+          LoopAPI.sendMessageToHandler({
+            name: "SetLoopPref",
+            data: [pref, timestamp]
+          });
+        }
+      });
```
Comment (mancas) [inline add-on/chrome/modules/MozLoopAPI.jsm]: Should we move this code to a new private function?

File: add-on/chrome/modules/MozLoopAPI.jsm [hunk 4]
```diff
@@ -1001,7 +1002,48 @@ const kMessageHandlers = {
    */
   TelemetryAddValue: function(message, reply) {
     let [histogramId, value] = message.data;
-    Services.telemetry.getHistogramById(histogramId).add(value);
+    if (histogramId === "LOOP_MAU") {
+      let buckets;
+      let pref = "mau.";
+      this.GetAllConstants(null, result => buckets = result.LOOP_MAU_TYPE);
```
Comment (mikedeboer) [inline add-on/chrome/modules/MozLoopAPI.jsm]: Please change this to `buckets = LOOP_MAU_TYPE;` - the call to `GetAllConstants` is unnecessary.

File: add-on/chrome/modules/MozLoopAPI.jsm [hunk 5]
```diff
@@ -1001,7 +1002,48 @@ const kMessageHandlers = {
...
+      switch(value) {
```
Comment (mikedeboer) [inline add-on/chrome/modules/MozLoopAPI.jsm]: nit: missing space after `switch`

File: add-on/chrome/modules/MozLoopAPI.jsm [hunk 6]
```diff
@@ -1001,7 +1002,48 @@ const kMessageHandlers = {
...
+          break;
+        case buckets.ROOM_SHARE:
+          pref += "roomShare";
+          break;
+        case buckets.ROOM_DELETE:
+          pref += "roomDelete";
+          break;
+      }
+
+      LoopAPI.sendMessageToHandler({
```
Comment (mikedeboer) [inline add-on/chrome/modules/MozLoopAPI.jsm]: let prefDate = MozLoopService.getLoopPref(pref) \* 1000;
let delta = Date.now() - prefDate;

File: add-on/chrome/modules/MozLoopAPI.jsm [hunk 7]
```diff
@@ -1001,7 +1002,48 @@ const kMessageHandlers = {
...
+        name: "GetLoopPref",
+        data: [pref]
+      }, result => {
+        var delta = new Date() - new Date(pref);
```
Comment (mikedeboer) [inline add-on/chrome/modules/MozLoopAPI.jsm]: This never worked, did it?

File: add-on/chrome/modules/MozLoopAPI.jsm [hunk 8]
```diff
@@ -1001,7 +1002,48 @@ const kMessageHandlers = {
...
+        // Send telemetry event if period (30 days) passed.
+        // 0 is default value for pref.
+        // 2592000 seconds in 30 days
+        if (pref === 0 || delta >= 2592000 * 1000) {
+          Services.telemetry.getHistogramById(histogramId).add(value);
+          let timestamp = Math.floor(new Date().getTime() / 1000);
+          LoopAPI.sendMessageToHandler({
```
Comment (mikedeboer) [inline add-on/chrome/modules/MozLoopAPI.jsm]: Please change this to `MozLoopService.setLoopPref(pref, Math.floor(Date.now() / 1000));`

File: add-on/chrome/modules/MozLoopAPI.jsm (hunk 3 above)
Comment (mikedeboer) [inline add-on/chrome/modules/MozLoopAPI.jsm]: See my comments above :smile:

File: add-on/chrome/modules/MozLoopAPI.jsm [hunk 9]
```diff
@@ -1001,7 +1002,48 @@ const kMessageHandlers = {
    */
   TelemetryAddValue: function(message, reply) {
     let [histogramId, value] = message.data;
-    Services.telemetry.getHistogramById(histogramId).add(value);
+    if (histogramId === "LOOP_MAU") {
+      let buckets;
+      let pref = "mau.";
```
Comment (mikedeboer) [inline add-on/chrome/modules/MozLoopAPI.jsm]: What if you did the following:

``` js
// Put this somewhere on top of the MessageHandlers code...
const kMauPrefMap = new Map(
  Object.getOwnPropertyNames(LOOP_MAU_TYPE).map(name => {
    let parts = name.toLowerCase().split("_");
    return [LOOP_MAU_TYPE[name], parts[0] + parts[1].charAt(0).toUpperCase() + parts[1].substr(1)];
  })
);
```

And here you can do:

``` js
let pref = "mau." + kMauPrefMap.get(value);
```

File: add-on/chrome/modules/MozLoopAPI.jsm (hunk 3 above)
Comment (mikedeboer) [inline add-on/chrome/modules/MozLoopAPI.jsm]: I'd also really appreciate if you'd put the `browser_mozLoop_telemetry.js` changes in this PR... even though you can't test them here :wink:

File: add-on/chrome/modules/MozLoopAPI.jsm (hunk 9 above)
Comment (mancas) [inline add-on/chrome/modules/MozLoopAPI.jsm]: Good idea!

File: add-on/chrome/modules/MozLoopAPI.jsm [hunk 10]
```diff
@@ -1001,7 +1008,22 @@ const kMessageHandlers = {
...
+    if (histogramId === "LOOP_MAU") {
+
+      let pref = "mau." + kMauPrefMap.get(value);
+      let prefDate = MozLoopService.getLoopPref(pref) * 1000;
+      let delta = Date.now() - prefDate;
+
+      // Send telemetry event if period (30 days) passed.
+      // 0 is default value for pref.
+      // 2592000 seconds in 30 days
+      if (pref === 0 || delta >= 2592000 * 1000) {
```
Comment (mikedeboer) [inline add-on/chrome/modules/MozLoopAPI.jsm:1045]: I'd really like to see this logic captured in a unit test...

File: add-on/chrome/modules/MozLoopAPI.jsm [hunk 11]
```diff
@@ -997,15 +1004,34 @@ const kMessageHandlers = {
    *                                                   in the histogram.
    *                           ]
    * @param {Function} reply   Callback function, invoked with the result of this
-   *                           message handler. The result will be sent back to
+   *                           message handlerg. The result will be sent back to
```
Comment (mikedeboer) [inline add-on/chrome/modules/MozLoopAPI.jsm]: I don't think you want this change :wink:

**Comment by mikedeboer** [comment]
Well done!! Don't forget the ESLint error.

**Comment by mikedeboer** [comment]
Great job with the tests, btw.

**Pull request merged**